source/x509storeissuer.c

```diff
@@ -20,158 +25,979 @@
+            timeout_us = parse_timeout(optarg);
+            break;
+        case 'n': /* nonce */
+            parse_nonce_cfg(optarg, &nonce_cfg);
+            break;
+        case 'm': /* mode */
+            if (strcasecmp(optarg, "r") == 0) {
+                mode = MODE_R;
+            } else if (strcasecmp(optarg, "rw") == 0) {
+                mode = MODE_RW;
+            } else {
+                errx(EXIT_FAILURE, "Unknown mode: \"%s\"", optarg);
+            }
+            break;
+        case 'w': /* maximum writers */
+            max_writers = parse_int(optarg, 0, INT_MAX,
+                                    "maximum number of writers");
+            break;
+        case 'W': /* percent of writes */
+            w_probability = (size_t) (parse_probability(optarg) * 65536 / 100);
+            break;
+        case 'l': /* max certs to load */
+            max_load_certs = parse_int(optarg, 0, INT_MAX,
+                                       "maximum certificate load count");
+            break;
+        case 'L': /* max certs dirs to load*/
+            max_load_cert_dirs = parse_int(optarg, 0, INT_MAX,
+                                           "maximum certificate directories"
+                                           " load count");
+            break;
+        case 'E':
+            skip_default_paths = true;
+            break;
+        case 'C': /* how many threads share X509_STORE_CTX */
+            ctx_share_cnt = parse_int(optarg, 1, INT_MAX,
+                                      "X509_STORE_CTX share degree");
             break;
         case 'V':
             perflib_print_version(basename(argv[0]));
             return EXIT_SUCCESS;
         default:
-            printf("Usage: %s [-t] [-V] certsdir threadcount\n", basename(argv[0]));
-            printf("-t - terse output\n");
-            printf("-V - print version information and exit\n");
+            usage(argv);
             return EXIT_FAILURE;
         }
     }
 
-    if (argv[optind] == NULL) {
-        printf("certsdir is missing\n");
-        goto err;
-    }
-    cert = perflib_mk_file_path(argv[optind], "servercert.pem");
-    if (cert == NULL) {
-        printf("Failed to allocate cert\n");
-        goto err;
-    }
-    optind++;
+    if (verbosity < VERBOSITY_VERBOSE)
+        quantiles = 1;
 
-    if (argv[optind] == NULL) {
-        printf("threadcount is missing\n");
-        goto err;
-    }
-    threadcount = atoi(argv[optind]);
-    if (threadcount < 1) {
-        printf("threadcount must be > 0\n");
-        goto err;
+    if (argv[optind] == NULL)
+        errx(EXIT_FAILURE, "certsdir is missing");
+
+    dirs_start = optind++;
+
+    /*
+     * Store the part of argv containing directories to nonce_cfg so
+     * load_nonce_from_path can use it later.
+     */
+    nonce_cfg.dirs = argv + dirs_start;
+    nonce_cfg.num_dirs = argc - 1 - dirs_start;
+
+    if (optind >= argc)
+        errx(EXIT_FAILURE, "threadcount is missing");
```

Comment:
IIRC all of these calls to errx exit immediately, instead of jumping to the err: label to free all the potentially allocated resources, resulting in leaks.  I don't know that anyone is going to care about leaks in the perf tests, but it seems like something we took care to avoid in the past

Reply:
I would not bother about doing proper release here. the process is going to die and OS will reclaim all memory, descriptors etc... so this is fine. If this approach sounds invalid because it may confuse some memory/resource checker then IMO it's better to fix the checker tool so it understands the err*()/exit() as ultimate resource free.